grpc/src/test/java/com/linecorp/armeria/client/grpc/GrpcClientTimeoutTest.java

```diff
@@ -0,0 +1,145 @@
+import static org.mockito.Mockito.atLeastOnce;
+import static org.mockito.Mockito.verify;
+
+import java.util.concurrent.TimeUnit;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.util.concurrent.Uninterruptibles;
+
+import com.linecorp.armeria.client.Clients;
+import com.linecorp.armeria.common.grpc.GrpcSerializationFormats;
+import com.linecorp.armeria.grpc.testing.Messages.SimpleRequest;
+import com.linecorp.armeria.grpc.testing.Messages.SimpleResponse;
+import com.linecorp.armeria.grpc.testing.TestServiceGrpc.TestServiceBlockingStub;
+import com.linecorp.armeria.grpc.testing.TestServiceGrpc.TestServiceImplBase;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.grpc.GrpcService;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.Appender;
+import io.grpc.Status.Code;
+import io.grpc.StatusRuntimeException;
+import io.grpc.stub.StreamObserver;
+
+class GrpcClientTimeoutTest {
+
+    private static final ch.qos.logback.classic.Logger rootLogger =
+            (ch.qos.logback.classic.Logger) LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME);
+
+    @Mock
+    private Appender<ILoggingEvent> appender;
+
+    @Captor
+    private ArgumentCaptor<ILoggingEvent> loggingEventCaptor;
+
+    @BeforeEach
+    void setupLogger() {
+        rootLogger.addAppender(appender);
+    }
+
+    @AfterEach
+    void cleanupLogger() {
+        rootLogger.detachAppender(appender);
+    }
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.requestTimeoutMillis(2000);
+            sb.service(GrpcService.builder()
+                                  .addService(new SlowService())
+                                  .build());
+        }
+    };
+
+    @Test
+    void clientTimeout() throws InterruptedException {
+        final TestServiceBlockingStub client =
+                Clients.newClient(server.httpUri(GrpcSerializationFormats.PROTO),
+                                  TestServiceBlockingStub.class);
+        final StatusRuntimeException exception = catchThrowableOfType(() -> {
+            client.withDeadlineAfter(1000, TimeUnit.MILLISECONDS)
+                  .unaryCall(SimpleRequest.getDefaultInstance());
+        }, StatusRuntimeException.class);
+        assertThat(exception.getStatus().getCode()).isEqualTo(Code.DEADLINE_EXCEEDED);
```

Comment:
If the understanding is correct,

Armeria gRPC client sets timeout when either option `Client.responseTimeout` or `io.grpc.Stub.withDelay` is specified.
After timed out, the client raises DEADLINE_EXCEEDED.

https://github.com/line/armeria/blob/armeria-1.2.0/grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/ArmeriaClientCall.java#L189-L207

Besides, if the either options of them is specified, the client send `grpc-timeout` to the server, and the server also set timeout.
After timed out at the server side, the server returns CANCELLED to the client.

https://github.com/line/armeria/blob/armeria-1.2.0/grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/ArmeriaClientCall.java#L398
https://github.com/line/armeria/blob/armeria-1.2.0/grpc/src/main/java/com/linecorp/armeria/server/grpc/FramedGrpcService.java#L190

So, I think there are two possibilities of the error code raised by the client.

Reply:
> If the understanding is correct,

Yes, exactly.

a gRPC client actively raises `DEADLINE_EXCEEDED` or `CANCELLED` by retruned status code.

Reply:
JFYI (I guess you already know this code), the following mapppings converter an exception to status code. 
https://github.com/line/armeria/blob/1b6d85fa0493f39c6d0b5b5084fe8440bf521223/grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/GrpcStatus.java#L78-L112

Reply:
In this test case, does the client never raise `CANCELLED`?

Reply:
Good point. It is possbile. I think it depends on Operating System scheduling. 
- A client sends a remaining timeout to a server. When the server schedules the timeout, network latency is not considered. Therefore, in most case, client will close a stream if a deadline exceeded.
 
- However, a client CPU is busy and hard to get CPU cycles to schedule a timeout for the deadline, a server timeout scheduler will be triggered and close a stream before the client deadline exceeded.

Reply:
Eventually to resolve this problem, as @trustin said in [slack channel](https://line-armeria.slack.com/archives/C1NGPBUH2/p1604287343245400?thread_ts=1604070308.229600&cid=C1NGPBUH2), we need to unify the timeout.